sensors: Driver for INA226

mayankmahajan-nxp opened the pull request:
INA226 - Bidirectional Current and Power Monitor w/ I2C
Boards Tested: mr_canhubk3

mayankmahajan-nxp commented:
Hi @MaureenHelm, could you please check this PR. Thank you.

mayankmahajan-nxp commented:
Hi @MaureenHelm, requesting a review on this PR. Kindly check at your convenience.

kartben commented on samples/sensor/ina226/README.rst:
please drop this sample entirely -- it is not exercising any API that would be specific to this sensor so is not adding value compared to e.g. sensor_shell sample. 
If you'd like to provide an example of how the binding can be used, you may add the devicetree snippet you have in your mr_canhubk3.overlay to the description field of the binding, as an example.

dleach02 replied:
@kartben is there a sample already in the tree where the author could add the board overlay to show it being tested?

kartben replied:
Like I said, `sensor_shell` sample should be enough for folks to exercise the sensor, since it only uses standard channels and API.
The overlay doesn't belong to code samples since AFAICT the mr_canhub overlay that was previously provided corresponded to a bespoke wiring (i.e. I don't think the mr_canhub comes with an INA226, please correct me if I'm wrong) so it is not really useful for folks. Also hence why I suggested to simply provide a "sample usage" in the description of the binding itself. Hope this helps.

kartben commented on drivers/sensor/ina226/ina226.h:
```diff
@@ -0,0 +1,54 @@
+#define INA226_REG_POWER		0x03
+#define INA226_REG_CURRENT		0x04
+#define INA226_REG_CALIB		0x05
+#define INA230_REG_MASK			0x06
+#define INA230_REG_ALERT		0x07
```
copy paste issue? prefix should probably be `INA226_`

mayankmahajan-nxp commented:
Hi @kartben, thank you for the review. I have addressed the comments. Kindly check at your convenience.

MaureenHelm commented on drivers/sensor/ina226/ina226.c:
```diff
@@ -0,0 +1,304 @@
+	uint16_t id;
+	int ret;
+
+	if (!device_is_ready(config->bus.bus)) {
```
Use `i2c_is_ready_dt`

MaureenHelm commented on dts/bindings/sensor/ti,ina226.yaml:
```diff
@@ -0,0 +1,103 @@
+properties:
+  config:
+    type: int
+    default: 0x0000
```
Is this the power-on reset value? Please explain in the description

MaureenHelm reviewed (changes requested):
Please add this sensor to `tests/drivers/build_all/sensor`

mayankmahajan-nxp commented:
> Please add this sensor to `tests/drivers/build_all/sensor`

Hi @MaureenHelm, thank you for the review. I have addressed the comments. Kindly check at your convenience.

dleach02 commented:
@mayankmahajan-nxp, please clean up the merge conflicts.

@MaureenHelm, @kartben, has the author addressed your review comments?

MaureenHelm commented on dts/bindings/sensor/ti,ina226.yaml:
```diff
@@ -0,0 +1,118 @@
+        the respective properties of the individual fields because we take the
+        bitwise OR of it and the dependent properties at device init.
+
+      Default is 0x0000 because it's only meant to control bit 15 (reset bit).
+        All other components of this property are meant to be assigned by
+        setting their respective property. e.g. for setting bits 2 - 0, we
+        should set the "operating-mode" property.
+        Note: The power-on reset value of configuration register is 0x4127.
```
Why do we need this property at all then? We shouldn't have redundant properties

mayankmahajan-nxp replied:
Yes that's correct. I have removed this property in the latest push. Sorry I didn't see it before.

MaureenHelm reviewed (commented):
Needs to be rebased

mayankmahajan-nxp commented:
Hi @dleach02 and @MaureenHelm, thank you for the review. I have resolved the merge conflicts and review comments. Kindly check at your convenience.

mayankmahajan-nxp commented:
Hi @kartben, requesting your review on this PR. Kindly check at your convenience. Thank you.

decsny commented on dts/bindings/sensor/ti,ina226.yaml:
```diff
@@ -0,0 +1,93 @@
+    type: int
+    required: true
+    description: |
+      This value should be selected so that measurement resolution is
```
this description does not make clear what the value means

mayankmahajan-nxp replied:
Hi @decsny, I have addressed this in the latest push. Kindly check if the description is clear now. If not then please let me know. Thank you.

mayankmahajan-nxp commented:
Hi @decsny, thank you for the review. I have addressed the comments. Kindly check at your convenience.
Hi @MaureenHelm, your approval was dismissed due to force push for the latest review. Kindly check. Thank you.

decsny commented:
added dev review label because kartben has not return with a block after some weeks

kartben commented:
> added dev review label because kartben has not return with a block after some weeks

I missed the notifications due to travelling - sorry. FWIW this kind of stale review is OK to dismiss without going to dev review, as it was clear my comments had been addressed (and thanks for that!)
I just realized I had another comment which I just posted, hopefully it can be addressed but happy if this is done in a follow-up commit if you don't want to lose the existing approvals

mayankmahajan-nxp commented:
Hi @kartben, thank you for the review. I have addressed the comments. Kindly check at your convenience. Thank you.
@MaureenHelm, @decsny Kindly check if you could approve this PR. I have addressed your comments. Thank you.

kartben commented on dts/bindings/sensor/ti,ina226.yaml:
```diff
@@ -0,0 +1,73 @@
+description: |
+    TI INA226 Bidirectional Current and Power Monitor.
+    The <zephyr/dt-bindings/sensor/ina226.h> file should be included in the
+    DeviceTree and it provides macros that can be used for initializing the
```
missed this small one before - should there be another round of updates: 
```suggestion
    DeviceTree as it provides macros that can be used for initializing the
```

kartben reviewed (dismissed):
@mayankmahajan-nxp  thanks!
Sorry if I wasn't clear, but keeping the binding sample was fine, I was only suggesting to have it focus on the "useful" bits. But this works with dropping it altogether too :)

mayankmahajan-nxp commented:
Hi @aescolar, I have rebased. Kindly check at your convenience. Thank you.

mayankmahajan-nxp committed:
ADD: Driver for Sensor INA226

INA226 - Bidirectional Current and Power Monitor w/ I2C
Boards Tested: mr_canhubk3

Signed-off-by: Mayank Mahajan <mayankmahajan.x@nxp.com>

mayankmahajan-nxp commented:
Hi, pushed again just now because a rebase was needed. @aescolar please check at your convenience. Thank you.
@MaureenHelm, @kartben, @decsny your approvals were dismissed due to force push. Kindly check at your convenience.

teburd commented on drivers/sensor/ti/ina226/ina226.h:
```diff
@@ -0,0 +1,50 @@
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/sensor.h>
+
+/* Device register addresses. */
```
There's no need to have a header as this is all local to the ina226.c compilation unit

mayankmahajan-nxp replied:
Ok @teburd. I could change it. But other INAXXX drivers (such as INA219) include a separate header. Kindly suggest how we should proceed.

mayankmahajan-nxp commented:
Hi @teburd, thank you for the review. I have addressed your comments. Kindly check at your convenience.
